lib/Doctrine/DBAL/Connections/PrimaryReplicaConnection.php

```diff
@@ -0,0 +1,361 @@
+        array $params,
+        Driver $driver,
+        ?Configuration $config = null,
+        ?EventManager $eventManager = null
+    ) {
+        if (! isset($params['replica'], $params['primary'])) {
+            throw new InvalidArgumentException('primary or replica configuration missing');
+        }
+
+        if (count($params['replica']) === 0) {
+            throw new InvalidArgumentException('You have to configure at least one replica.');
+        }
+
+        $params['primary']['driver'] = $params['driver'];
+        foreach ($params['replica'] as $replicaKey => $replica) {
+            $params['replica'][$replicaKey]['driver'] = $params['driver'];
+        }
+
+
+        parent::__construct($params, $driver, $config, $eventManager);
+    }
+
+    private function switchToPrimary(): bool
+    {
+        $forcePrimaryAsReplica = $this->getTransactionNestingLevel() > 0;
+
+        if (isset($this->primary)) {
+            $this->_conn = $this->primary;
+
+            if ($forcePrimaryAsReplica) {
+                $this->replica = $this->_conn;
+            }
+
+            return false;
+        }
+
+        $this->primary = $this->_conn = $this->connectTo('primary');
+
+        // Set replica connection to primary to avoid invalid reads
+        $this->replica = $this->primary;
+
+        if ($this->_eventManager->hasListeners(Events::postConnect)) {
+            $eventArgs = new ConnectionEventArgs($this);
+            $this->_eventManager->dispatchEvent(Events::postConnect, $eventArgs);
+        }
+
+        return true;
+    }
+
+    public function connect(): bool
+    {
+        $connectionName = 'replica';
+
+        if ($this->_conn !== null) {
+            return false;
+        }
+
+        $forcePrimaryAsReplica = false;
+
+        if ($this->getTransactionNestingLevel() > 0) {
+            $connectionName        = 'primary';
+            $forcePrimaryAsReplica = true;
+        }
+
+        if (isset($this->$connectionName)) {
+            $this->_conn = $this->$connectionName;
+
+            if ($forcePrimaryAsReplica) {
+                $this->replica = $this->_conn;
+            }
+
+            return false;
+        }
+
+        if ($connectionName === 'primary') {
+            $this->primary = $this->_conn = $this->connectTo($connectionName);
+
+            // Set replica connection to primary to avoid invalid reads
+            $this->replica = $this->primary;
+        } else {
+            $this->replica = $this->_conn = $this->connectTo($connectionName);
+        }
+
+        if ($this->_eventManager->hasListeners(Events::postConnect)) {
+            $eventArgs = new ConnectionEventArgs($this);
+            $this->_eventManager->dispatchEvent(Events::postConnect, $eventArgs);
+        }
+
+        return true;
+    }
+
+    /**
+     * Connects to a specific connection.
+     */
+    private function connectTo(string $connectionName): DriverConnection
+    {
+        $params = $this->getParams();
```

Comment:
This is BTW one of the places where I couldn't get rid of the call to `getParams()` (https://github.com/doctrine/dbal/issues/3593). The purpose of the connection is to transmit the data between the client and the server, not expose the connection parameters via the public API.

Having these implementation details exposed as is via the API makes refactoring of the internals close to impossible.